# Introduction

## Overview
This notebook preprocesses the original event dataset by refining its structure and extracting the most relevant features for our experiments. The preprocessing ensures that the data is suitable for analysis. Additionally, the notebook creates manageable samples from the dataset, allowing for experiments on a feasible number of instances. This preprocessing step lays the foundation for subsequent analysis and experimentation.

## Purpose
The dataset requires preprocessing to prepare it for analysis. The key steps include:

* Feature Selection – Filtering the most relevant features, focusing on the semantic, geospatial, and temporal aspects of an event (see README.md for details).
* Handling Missing Values – Removing events with NaN values in the selected features.
* Formatting Features – Standardizing unformatted data to ensure consistency.

We also sample the original dataset, obtaining medium-sized datasets to enable later experimentation with less computationally efficient methods.

# Outset Steps

### Reading parameters

Bellow, a parameter (see `run_experiments.sh` and `preprocessing.sh`) for the output directory, where the sampled datasets will be stored.

In [ ]:
output_dir = None

In [ ]:
if output_dir is None:
    output_dir = "./UsageDatasets/"

### Downloading data

Getting the data from google drive (can run for a little long, data has 2,4GB).

In [1]:
!gdown --quiet 104WLoNAy02-r7d18ZE16vudY0puVpxm2

### Loading relevant packages and setting logging

In [2]:
import sys
import logging
from pathlib import Path

import pandas as pd

In [3]:
log_filename = "../logs/pre_processing.log"
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(levelname)s - %(message)s",
    handlers=[
        logging.FileHandler(log_filename, mode="w"),  # Overwrite log file
        logging.StreamHandler(sys.stdout)  # Print to console
    ]
)

# Data Preprocessing

In the following cells, we load the Dataset into a Pandas DataFrame. One may want to inspect the dataset to better understand our data in the associated processes. Remember events come from a news article.

In [4]:
df_event_data = pd.read_pickle("df_event_data_with_embeddings.pkl") 

In [5]:
df_event_data 

,gid,url,geoinfo1,taxon1,taxon2,geoinfo2,who1,who2,text,predict,predict_proba,country,country_code,country_lat,country_lng,embeddings
924124,20200214223000-56,https://www.hppr.org/post/how-covid-19-kills-n...,,KILL;CRISISLEX_T03_DEAD;TAX_DISEASE;TAX_DISEAS...,"CRISISLEX_CRISISLEXREC,3366;TAX_DISEASE_INFECT...","4#Wuhan, Hubei, China#CH#CH12#13124#30.5833#11...","Anthony Fauci,3509;Yoko Furuya,1354;Sylvie Bri...","Columbia University Irving Medical Center,1434...",How COVID-19 Kills: The New Coronavirus Diseas...,covid-19 vaccine,1.000000,"Wuhan, Hubei, China",CH,30.5833,114.267,"[-0.020972056, -0.12172927, -0.007201001, -0.0..."
2284004,20200228053000-1137,http://www.ecns.cn/news/politics/2020-02-28/de...,KILL#13##1#South Korea#KS#KS#37#127.5#KS#2101;...,TAX_ETHNICITY;TAX_ETHNICITY_CHINESE;TAX_WORLDL...,"TAX_DISEASE_EPIDEMIC,1288;TAX_DISEASE_EPIDEMIC...",1#South Korea#KS#KS##37#127.5#KS#33;1#South Ko...,"Xing Haiming,46;Xing Haiming,240;Xing Haiming,...","Chinese Embassy,157;Chinese Embassy,351;Chines...",Chinese Embassy in South Korea donates masks t...,funding for covid,1.000000,South Korea,KS,37,127.5,"[0.05996564, -0.16531578, -0.0031238094, 0.057..."
1181064,20200224034500-218,https://www.rnz.co.nz/news/national/410233/cov...,,TAX_FNCACT;TAX_FNCACT_MINISTER;LEADER;TAX_FNCA...,"CRISISLEX_CRISISLEXREC,1466;UNGP_HEALTHCARE,74...",1#South Korea#KS#KS##37#127.5#KS#1508;1#Chines...,"Jacinda Arden,28","World Health Organisation,1182",Covid-19: Travel restrictions for those coming...,covid-19 restrictions lifted,1.000000,South Korea,KS,37,127.5,"[-0.104065694, 0.03418176, -0.0051460676, -0.0..."
3572684,20200224134500-9,https://www.chinanationalnews.com/news/2641216...,KILL#150##1#China#CH#CH#35#105#CH#1919;KILL#2#...,TAX_ETHNICITY;TAX_ETHNICITY_INDIAN;TAX_FNCACT;...,"GENERAL_HEALTH,1852;MEDICAL,1852;GENERAL_GOVER...",1#South Korea#KS#KS##37#127.5#KS#18;1#South Ko...,,"South China Morning Post,1835;Yonhap News Agen...","COVID-19: Reconsider travelling to S Korea, In...",hospital for covid-19 patients,0.333333,South Korea,KS,37,127.5,"[-0.043811183, -0.0037411905, -0.005606968, 0...."
3188350,20200224071500-424,https://www.thenorthernecho.co.uk/news/nationa...,KILL#2##1#China#CH#CH#35#105#CH#134;KILL#8##1#...,KILL;TAX_FNCACT;TAX_FNCACT_OFFICIALS;TAX_DISEA...,"GENERAL_GOVERNMENT,2051;EPU_POLICY_GOVERNMENT,...","4#Milan, Lombardia, Italy#IT#IT09#18363#45.466...",,"Xinhua News Agency,1731;Xinhua,1719;Xinhua,1798",South Korea sees increase in recorded Covid-19...,covid-19 cases emerge,0.666667,"Milan, Lombardia, Italy",IT,45.4667,9.2,"[-0.026586916, -0.3584028, -0.0059149084, -0.0..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1029937,20210309233000-1970,https://www.tbnewswatch.com/local-news/thunder...,,SECURITY_SERVICES;TAX_FNCACT;TAX_FNCACT_POLICE...,"MEDICAL,768;MEDICAL,1151;TAX_FNCACT_PARAMEDICS...","4#Thunder Bay, Ontario, Canada#CA#CA08#12684#4...","Bill Mauro,1354;Kristen Oliver,153",,Thunder Bay frontline officers receive first d...,covid-19 vaccine,1.000000,"Thunder Bay, Ontario, Canada",CA,48.4,-89.2333,"[-0.074297085, -0.2056211, -0.0043920088, -0.1..."
180173,20210308140000-411,https://www.ross-shirejournal.co.uk/news/ross-...,,TAX_FNCACT;TAX_FNCACT_WOMEN;EPU_ECONOMY;EPU_EC...,"GENERAL_GOVERNMENT,1355;EPU_POLICY_GOVERNMENT,...",1#Scotland#UK#UK##54#-4#UK#428;1#Scotland#UK#U...,"Maree Todd,24","Scottish Government Women Returners Programme,...",Women have suffered 'disproportionately' from ...,impact of the covid-19,1.000000,Scotland,UK,54,-4,"[-0.10385181, 0.0008548315, -0.0063411416, -0...."
703959,20210302051500-673,https://www.republicworld.com/world-news/pakis...,KILL#896##1#Pakistan#PK#PK#30#70#PK#647;,GENERAL_HEALTH;MEDICAL;TAX_DISEASE;TAX_DISEASE...,"GENERAL_GOVERNMENT,1807;GENERAL_GOVERNMENT,216...",1#Pakistani#PK#PK##30#70#PK#141;1#Pakistan#PK#...,"Qaisar Sajjad,1518","Pakistan National,689;Pakistan Medical Associa...",Pakistan's NCOC raises alarm over dull vaccine...,covid-19 vaccine,1.0

Next, we select the relevant features for our analysis, focusing on the semantic, geospatial, and temporal characteristics of each event. The selected features are:

* _gid_ – Unformatted date of the event news **(yyyy-mm-dd-hh-mm)**.
* _predict_ – A label that will be useful for sampling.
* _country lat_ – Latitude of the event location.
* _country long_ – Longitude of the event location.
* _embeddings_ – The news article processed using a BERT model.

We've decided to save the embeddings directly instead of doing the encoding process here. Despite the memory demand, this eliminates one's need for a GPU when processing the dataset.

In [6]:
df_selected_features = df_event_data[["gid","predict","country_lat","country_lng","embeddings"]].copy(deep=True)

One could create a deep copy to avoid view-related warnings, but it's not necessary here since we're not modifying or reusing the source dataset.

In [7]:
df_selected_features

,gid,predict,country_lat,country_lng,embeddings
924124,20200214223000-56,covid-19 vaccine,30.5833,114.267,"[-0.020972056, -0.12172927, -0.007201001, -0.0..."
2284004,20200228053000-1137,funding for covid,37,127.5,"[0.05996564, -0.16531578, -0.0031238094, 0.057..."
1181064,20200224034500-218,covid-19 restrictions lifted,37,127.5,"[-0.104065694, 0.03418176, -0.0051460676, -0.0..."
3572684,20200224134500-9,hospital for covid-19 patients,37,127.5,"[-0.043811183, -0.0037411905, -0.005606968, 0...."
3188350,20200224071500-424,covid-19 cases emerge,45.4667,9.2,"[-0.026586916, -0.3584028, -0.0059149084, -0.0..."
...,...,...,...,...,...
1029937,20210309233000-1970,covid-19 vaccine,48.4,-89.2333,"[-0.074297085, -0.2056211, -0.0043920088, -0.1..."
180173,20210308140000-411,impact of the covid-19,54,-4,"[-0.10385181, 0.0008548315, -0.0063411416, -0...."
703959,20210302051500-673,covid-19 vaccine,30,70,"[0.018137967, 0.01642545, -0.0037686664, -0.07..."
2108475,20210520121500-755,covid-19 restrictions eased,40.3363,-89.0022,"[-0.07282667, 0.09720857, -0.0031480393, -0.07..."


# Data Cleaning & Feature Engineering 

Here, we check the data types and non-null counts of each relevant feature to guide the data cleaning and feature engineering process.

In [8]:
df_selected_features.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392105 entries, 924124 to 486543
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   gid          392105 non-null  object
 1   predict      392105 non-null  object
 2   country_lat  392105 non-null  object
 3   country_lng  392105 non-null  object
 4   embeddings   392105 non-null  object
dtypes: object(5)
memory usage: 17.9+ MB


The latitude and longitude coordinates are identified as `object`. We need to convert them to float type to enable future operations. 

In [9]:
df_selected_features.country_lat.head(15)

924124     30.5833
2284004         37
1181064         37
3572684         37
3188350    45.4667
399287          46
2818152           
3669290    42.1497
2494842    55.8667
3060487    22.2833
2292890    30.5833
1179353           
3376499    45.6989
1711272    30.5833
11980      30.5833
Name: country_lat, dtype: object

One can see that there are null values, assigned with ''. Fortunately, `pd.to_numeric` handles this gracefully.

In [10]:
df_selected_features['country_lat'] = pd.to_numeric(df_selected_features['country_lat'], errors='raise')
df_selected_features['country_lng'] = pd.to_numeric(df_selected_features['country_lng'], errors='raise')

In [11]:
df_selected_features.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392105 entries, 924124 to 486543
Data columns (total 5 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   gid          392105 non-null  object 
 1   predict      392105 non-null  object 
 2   country_lat  336195 non-null  float64
 3   country_lng  336195 non-null  float64
 4   embeddings   392105 non-null  object 
dtypes: float64(2), object(3)
memory usage: 17.9+ MB


One can easily see 55,910 instances with null coordinates. A more in-depth discussion on handling these missing values is beyond our scope here. One may notice we'll sample the dataset, as a matter of fact.

We drop the data without geoespatial coordinates.

In [12]:
df_selected_features = df_selected_features.dropna(subset={'country_lat','country_lng'})

Now, we need to get the temporal information we have about the events. This info is encoded gid and can be extracted as follows.


In [13]:
dates = [date[:4]+'-'+date[4:6]+'-'+date[6:8] + ' ' + date[8:10] + ':' + date[10:12]  for date in df_selected_features.gid]

df_selected_features.loc[:, 'dates'] = pd.to_datetime(dates, errors='raise')

/tmp/ipykernel_206148/2380539424.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected_features.loc[:, 'dates'] = pd.to_datetime(dates, errors='raise')


In [14]:
df_selected_features

,gid,predict,country_lat,country_lng,embeddings,dates
924124,20200214223000-56,covid-19 vaccine,30.5833,114.26700,"[-0.020972056, -0.12172927, -0.007201001, -0.0...",2020-02-14 22:30:00
2284004,20200228053000-1137,funding for covid,37.0000,127.50000,"[0.05996564, -0.16531578, -0.0031238094, 0.057...",2020-02-28 05:30:00
1181064,20200224034500-218,covid-19 restrictions lifted,37.0000,127.50000,"[-0.104065694, 0.03418176, -0.0051460676, -0.0...",2020-02-24 03:45:00
3572684,20200224134500-9,hospital for covid-19 patients,37.0000,127.50000,"[-0.043811183, -0.0037411905, -0.005606968, 0....",2020-02-24 13:45:00
3188350,20200224071500-424,covid-19 cases emerge,45.4667,9.20000,"[-0.026586916, -0.3584028, -0.0059149084, -0.0...",2020-02-24 07:15:00
...,...,...,...,...,...,...
1029937,20210309233000-1970,covid-19 vaccine,48.4000,-89.23330,"[-0.074297085, -0.2056211, -0.0043920088, -0.1...",2021-03-09 23:30:00
180173,20210308140000-411,impact of the covid-19,54.0000,-4.00000,"[-0.10385181, 0.0008548315, -0.0063411416, -0....",2021-03-08 14:00:00
703959,20210302051500-673,covid-19 vaccine,30.0000,70.00000,"[0.018137967, 0.01642545, -0.0037686664, -0.07...",2021-03-02 05:15:00
2108475,20210520121500-755,covid-19 restrictions eased,40.3363,-89.00220,"[-0.07282667, 0.09720857, -0.0031480393, -0.07...",2021-05-20 12:15:00


The resultant dataset got what we need to perform our experiments.

In [15]:
df_processed = df_selected_features[['predict', 'country_lat', 'country_lng', 'embeddings','dates']]

In [16]:
df_processed

,predict,country_lat,country_lng,embeddings,dates
924124,covid-19 vaccine,30.5833,114.26700,"[-0.020972056, -0.12172927, -0.007201001, -0.0...",2020-02-14 22:30:00
2284004,funding for covid,37.0000,127.50000,"[0.05996564, -0.16531578, -0.0031238094, 0.057...",2020-02-28 05:30:00
1181064,covid-19 restrictions lifted,37.0000,127.50000,"[-0.104065694, 0.03418176, -0.0051460676, -0.0...",2020-02-24 03:45:00
3572684,hospital for covid-19 patients,37.0000,127.50000,"[-0.043811183, -0.0037411905, -0.005606968, 0....",2020-02-24 13:45:00
3188350,covid-19 cases emerge,45.4667,9.20000,"[-0.026586916, -0.3584028, -0.0059149084, -0.0...",2020-02-24 07:15:00
...,...,...,...,...,...
1029937,covid-19 vaccine,48.4000,-89.23330,"[-0.074297085, -0.2056211, -0.0043920088, -0.1...",2021-03-09 23:30:00
180173,impact of the covid-19,54.0000,-4.00000,"[-0.10385181, 0.0008548315, -0.0063411416, -0....",2021-03-08 14:00:00
703959,covid-19 vaccine,30.0000,70.00000,"[0.018137967, 0.01642545, -0.0037686664, -0.07...",2021-03-02 05:15:00
2108475,covid-19 restrictions eased,40.3363,-89.00220,"[-0.07282667, 0.09720857, -0.0031480393, -0.07...",2021-05-20 12:15:00


# Creating Final Datasets

In this section, we present the code employed to sample the processed original data and create datasets based on the original labels. Preserving the labels allows us to establish a criterion for sampling the data.

### Utility functions

We define two functions: one to extract and format the label for creating sampled datasets, and another to generate medium-sized datasets from these samples.

In [17]:
def formatting_label_name(str: str) -> str:
  return str.title().replace(" ","")

def create_datasets(data: pd.DataFrame, label: str, num_rows: int, random_state: int = 42) -> pd.DataFrame:
    dataset = data[data['predict'] == label]
    return dataset.sample(num_rows, random_state=random_state) if len(dataset) > num_rows else dataset

### Sampling 

We must take a look on the amount of events associated to each label.

In [18]:
df_processed['predict'].value_counts()

predict
covid-19 vaccine                   69975
 warned of covid-19                60502
 new covid-19 cases                31455
 deaths from covid-19              29308
 hospital for covid-19 patients    27199
covid-19 cases emerge              26270
 funding for covid                 23751
 tests for covid-19                21432
 covid-19 restrictions lifted       9182
 drug against covid-19              7302
 impact of the covid-19             6684
 covid-19 restrictions eased        5287
 treatments for covid-19            4130
 response to covid-19               2672
 reopening schools                  2087
 resurgence of covid-19             1592
 market analysis                    1185
 enforcing covid-19 protocols       1123
 breach covid rules                 1078
 festival cancelled                  997
 wears face mask                     898
 social distancing                   557
 quarantine for infected             470
 donate blood plasma                 387
research

For analysis purposes, it's interesting to have some variability on the datasets sizes. We can choose the following labels:
<table border="1">
  <thead>
    <tr>
      <th>Label</th>
      <th>Instances</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>covid-19 vaccine</td>
      <td>69975</td>
    </tr>
    <tr>
      <td>warned of covid-19</td>
      <td>60502</td>
    </tr>
      <td> new covid-19 cases</td>
      <td>31455</td>
    </tr>
    <tr>
      <td>tests for covid-19</td>
      <td>21432</td>
    </tr>
    <tr>
      <td>covid-19 restrictions lifted</td>
      <td>9182</td>
    </tr>
    <tr>
      <td>covid-19 restrictions eased</td>
      <td>5287</td>
    </tr>
  </tbody>
</table>

Because there's an interesting original size gap between them. However, we must sample them to get finner numbers and make the experiments more feasiable. The propposed samples sizes are as follows:

<table border="1">
  <thead>
    <tr>
      <th>Label</th>
      <th>Instances</th>
      <th>Sample size</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>covid-19 vaccine</td>
      <td>69975</td>
      <td>22500</td>
    </tr>
    <tr>
      <td>warned of covid-19</td>
      <td>60502</td>
      <td>16900</td>
    </tr>
      <td> new covid-19 cases</td>
      <td>31455</td>
      <td>12100</td>
    </tr>
    <tr>
      <td>tests for covid-19</td>
      <td>21432</td>
      <td>10000</td>
    </tr>
    <tr>
      <td>covid-19 restrictions lifted</td>
      <td>9182</td>
      <td>4900</td>
    </tr>
    <tr>
      <td>covid-19 restrictions eased</td>
      <td>5287</td>
      <td>2500</td>
    </tr>
  </tbody>
</table>


A more comprehensive discussion of this sampling methodology will be provided at an appropriate time, specifically in the published article.

In [19]:
samples = [
    ("covid-19 vaccine", 22500),
    (" warned of covid-19", 16900),
    (" new covid-19 cases", 12100),
    (" tests for covid-19", 10000),
    (" covid-19 restrictions lifted", 4900),
    (" covid-19 restrictions eased", 2500)
]


## Saving pre-processed data

We finally sample the data, creating pickles for the resultant datasets.

In [21]:
for sample in samples:
  label, size = sample
  dataset = create_datasets(data = df_processed, label = label, num_rows = size)
  print(dataset.shape)
  dataset.to_pickle(f"{output_dir}/{formatting_label_name(label)}_dataset.pkl")

(22500, 5)


(16900, 5)


(12100, 5)


(10000, 5)


(4900, 5)


(2500, 5)


In [22]:
logging.info("Preprocessing completed successfully")

2025-03-19 22:51:54,420 - INFO - Preprocessing completed successfully
